### ISSUE TITLE
Current algorithm is racy and suboptimal

### ISSUE THREAD

**Issue by domenic** [WICG/origin-isolation#6]
The algorithm in the current [specification plan](https://github.com/domenic/origin-isolation/tree/40c6de6a3143cf73cd710edfa62148cfe3479670#specification-plan) is suboptimal in scenarios such as:

- Main frame `https://example.com/` requests OI
- Sub frame A `https://sub.example.com/` does not request OI
- Sub frame B `https://sub2.example.com/` requests OI

In this case whether sub frame B gets origin-isolated or not depends on whether it loads first.

This is pretty bad and we should do better. As a first step we're trying to compile a list of scenarios. Currently we're doing this work [in a Google doc](https://docs.google.com/document/d/1LRi-oUYbWyXKUXek7WtZ3Pxgg4Etf1VewpqSnVIN0Do/edit#) for easier collaborative editing; we will move it to GitHub soon. This is a spec tracking issue and a heads-up to anyone watching.

**Comment by annevk** [comment]
A workaround might be that a "site agent cluster" also tracks the origins that belong to it. And only if both match you get to join it directly, otherwise your OI status is consulted.

That does not address the race between same-site-but-cross-origin subframes of the same origin, however. It seems impossible to avoid that race except if you let the top-level decide for the entire browsing context group as per COOP+COEP.

(Note that the scenarios apply equally to popups. It doesn't really alter anything here I think, but seems worth noting it's not just subframes.)

**Comment by domenic** [comment]
> That does not address the race between same-site-but-cross-origin subframes of the same origin, however. It seems impossible to avoid that race except if you let the top-level decide for the entire browsing context group as per COOP+COEP.

Is that the following scenario?

- Main frame `https://example.com/` requests OI
- Sub frame A `https://sub.example.com/` does not request OI
- Sub frame B `https://sub.example.com/` requests OI

I think it should be somehow possible to allocate these as main = `Origin{https://example.com/}`, A = `Site{https://example.com/}`, B = `Origin{https://sub.example.com/}`, because there's no way for A and B to get a reference to each other.

The trick is coming up with an algorithm that does this...

**Comment by domenic** [comment]
Nevermind, A and B can get a reference to each other via `top[0]` and `top[1]`. OK, I do agree the race there seems unavoidable.

By the way I think this is doable without explicitly tracking the origins, at least in spec land. You can check if the BCG's agent cluster map contains _site_ as a key, and bcg's agent cluster map[_site_] contains any agents which contain any realms whose relevant settings object's origins are same-origin with origin. In other words, the spec already has a way to get a list of all origins in an agent cluster.

**Comment by annevk** [comment]
Wouldn't that mean that as you go back and forward in history (without bfcache) you could end up in different agent clusters?

**Comment by domenic** [comment]
Isn't that already the case? E.g. if I navigate from `https://example.com/` to `https://example.org/` the second document's realm's agent will be in a different agent cluster compared to the first. And then we go back to the first agent cluster if you press the back button.

**Comment by annevk** [comment]
Why? What removes the agent cluster from the browsing context that navigation occurs in?

**Comment by domenic** [comment]
Nothing removes the agent cluster, but the document gets put in a second, new agent cluster.

**Comment by annevk** [comment]
Where is that defined?

I think that's possible to allow if the BCG contains a single TLBC, but it's not defined by the specification today, even if it's not observable in the single TLBC case as there's no origin isolation yet. (We don't define replacing BCGs yet, that's waiting on COOP.)

**Comment by domenic** [comment]
I don't think it's defined; we have a big XXX box around navigation and agents in the spec.

However, we certainly are not going to put cross-site documents (my `https://example.com/` and `https://example.org/`) in the same agent cluster in the implementation, just for security purposes.

**Comment by annevk** [comment]
Oh sorry, yeah, I misunderstood. My [earlier comment](https://github.com/domenic/origin-isolation/issues/6#issuecomment-572448879) was about same-site navigations.

**Comment by domenic** [comment]
Let me open a new issue to capture the remaining concerns since this one got a bit sidetracked and the original issue is solved.
